File: 138-复制带随机指针的链表.cpp
class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(!head)
            return head;

        unordered_map<Node*, Node*> mymap;
        Node* new_head = new Node(head->val);
        mymap[head] = new_head;
        mymap[NULL] = NULL;
        
        while(head){
            if(mymap.find(head->random) == mymap.end())
                mymap[head->random] = new Node(head->random->val);
            mymap[head]->random = mymap[head->random];

            if(mymap.find(head->next) == mymap.end())
                mymap[head->next] = new Node(head->next->val);
            mymap[head]->next = mymap[head->next];

            head = head->next;
        }

        return new_head;
    }
};

File: 138-复制带随机指针的链表.py
class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head:
            return None

        mylist = []
        mydict = {None: None};
        cur_node = head
        i = 0

        while cur_node:
            mylist.append((Node(cur_node.val), cur_node.random))
            mydict[cur_node] = i
            cur_node = cur_node.next
            i += 1
        
        for i in range(len(mylist) - 1):
            node1, node2 = mylist[i]
            node1.next = mylist[i + 1][0]
            node1.random = mylist[mydict[node2]][0] if mydict[node2] is not None else None

        node1, node2 = mylist[-1]
        node1.random = mylist[mydict[node2]][0] if mydict[node2] is not None else None

        return mylist[0][0]
